File: java/adaptersdk-schema-java-generator/src/main/java/io/github/fileanalysissuite/adaptersdk/schema/generator/SchemaGeneratorHelper.java
/*
 * Copyright 2022 Micro Focus or one of its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.fileanalysissuite.adaptersdk.schema.generator;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

import org.apache.commons.io.IOUtils;
import org.codehaus.plexus.util.StringUtils;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;

final class SchemaGeneratorHelper
{
    static final String PACKAGE_NAME = "io.github.fileanalysissuite.adaptersdk.schema";
    static final String CLASS_NAME = "AdapterSdkSchema";
    static final Set<String> PROPERTY_TYPES
        = new HashSet<>(Arrays.asList(
            "STRING", "FULLTEXT", "LONG", "DOUBLE", "INTEGER", "BOOLEAN", "DATETIME",
            "STRING[]", "FULLTEXT[]", "LONG[]", "DOUBLE[]", "INTEGER[]", "BOOLEAN[]", "DATETIME[]"));

    private SchemaGeneratorHelper() {}

    static JsonNode parseSchemaDefinition()
    {
        try {
            return parseSchema("schemaDefinition.json.yaml");
        } catch (final IOException e) {
            throw new RuntimeException("Unable to parse schema", e);
        }
    }

    static JsonNode parseSchema(final String schemaPath) throws IOException
    {
        final String schema;
        try (final InputStream stream = SchemaObjectBuilderGenerator.class.getResourceAsStream(schemaPath)) {
            schema = IOUtils.toString(stream, StandardCharsets.UTF_8);
        } catch (final IOException e) {
            throw new RuntimeException("Error reading schema definition from " + schemaPath, e);
        }
        return new ObjectMapper(new YAMLFactory()).readTree(schema);
    }

    static String toProperCase(final String str)
    {
        return StringUtils
            .capitaliseAllWords(str.replace('_', ' ').replace('-', ' ').toLowerCase(Locale.US))
            .replace(" ", "");
    }

    static String toFieldNameCase(final String str)
    {
        return StringUtils.lowercaseFirstLetter(toProperCase(str));
    }

    static String[] addExtraElement(final String[] element, final String extraElement)
    {
        final String[] newPath = Arrays.copyOf(element, element.length + 1);
        newPath[element.length] = extraElement;
        return newPath;
    }
}

File: dotnet/src/OpenText.CDDRI.AdapterSdkSchema.Generator/SchemaGeneratorHelper.cs
/**
 * Copyright 2022-2025 Open Text.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text.Json.Nodes;
using YamlDotNet.Serialization;

namespace OpenText.CDDRI.AdapterSdkSchema
{
    internal static class SchemaGeneratorHelper
    {
        internal static readonly string NAMESPACE_NAME = "OpenText.CDDRI.AdapterSdkSchema";
        internal static readonly string CLASS_NAME = "AdapterSdkSchema";
        internal static readonly HashSet<string> PROPERTY_TYPES = new() {
            "STRING", "FULLTEXT", "LONG", "DOUBLE", "INTEGER", "BOOLEAN", "DATETIME",
            "STRING[]", "FULLTEXT[]", "LONG[]", "DOUBLE[]", "INTEGER[]", "BOOLEAN[]", "DATETIME[]"};
        internal static readonly string[] PRIMITIVE_NULLABLE_TYPES = new[] { "LONG","DOUBLE","INTEGER","BOOLEAN"};

        internal static JsonNode ParseSchemaDefinition()
        {
            return ParseSchema("schemaDefinition.json.yaml");
        }

        internal static JsonNode ParseSchema(string schemaPath)
        {
            var assembly = Assembly.GetExecutingAssembly();
            string schemaResourceName = assembly.GetManifestResourceNames().Single(str => str.EndsWith(schemaPath));

            using Stream schemaDefinitionStream = assembly.GetManifestResourceStream(schemaResourceName);
            using StreamReader schemaReader = new(schemaDefinitionStream);

            var schemaDeserializer = new Deserializer();
            var yamlSchema = schemaDeserializer.Deserialize(schemaReader);

            var serializer = new SerializerBuilder().JsonCompatible().Build();
            var json = serializer.Serialize(yamlSchema);

            return JsonNode.Parse(json);
        }

        internal static string ToProperCase(string str)
        {
            TextInfo textInfo = new CultureInfo("en-US", false).TextInfo;
            return textInfo
                .ToTitleCase(str.Replace('_', ' ').Replace('-', ' ').ToLower())
                .Replace(" ", "");
        }

        internal static string ToFieldNameCase(string str)
        {
            string properCase = ToProperCase(str);
            return properCase.Length > 1
                ? char.ToLowerInvariant(properCase[0]) + properCase.Substring(1)
                : char.ToLowerInvariant(properCase[0]).ToString();
        }

        internal static string[] AddExtraElement(string[] element, string extraElement)
        {
            string[] newPath = new string[element.Length + 1];
            element.CopyTo(newPath, 0);
            newPath[element.Length] = extraElement;
            return newPath;
        }

        internal static string ToValidatorFieldName(string propertyName)
        {
            return "_is" + SchemaGeneratorHelper.ToProperCase(propertyName) + "Set";
        }

        internal static bool IsPrimitiveNullableType(string fieldTypeValue)
        {
            return PRIMITIVE_NULLABLE_TYPES.Contains(fieldTypeValue);
        }
    }
}
